EchoBot.cs

```diff
@@ -0,0 +1,177 @@
+            JObject headerObject = JObject.Parse(Base64UrlEncoder.Decode(jwtSecurityToken.RawHeader));
+
+            string nonceIn = headerObject.GetValue("nonce", StringComparison.OrdinalIgnoreCase).ToString();
+            string hashName = headerObject.GetValue("alg", StringComparison.OrdinalIgnoreCase).ToString();
+
+            HashAlgorithm hashAlgorithm = EchoBot.GetHashFunction(hashName);
+            string nonceOut = Base64UrlEncoder.Encode(hashAlgorithm.ComputeHash(Encoding.UTF8.GetBytes(nonceIn)));
+            headerObject["nonce"] = nonceOut;
+            string newHeaderString = Base64UrlEncoder.Encode(headerObject.ToString(Formatting.None));
+            string postTransformedToken = $"{newHeaderString}.{jwtSecurityToken.RawPayload}.{jwtSecurityToken.RawSignature}";
+            return postTransformedToken;
+        }
+
+        private static HashAlgorithm GetHashFunction(string hashAlgorithmName)
+        {
+            switch (hashAlgorithmName)
+            {
+                case "RS256":
+                    return SHA256.Create();
+                default:
+                    string errorMessage = $"Algorithm [{hashAlgorithmName}] not supported for PFT at this time";
+                    throw new NotSupportedException(errorMessage);
+            }
+        }
+
+        private async Task<string> getActorToken()
+        {
+            string url = "https://login.microsoftonline.com/f8cdef31-a31e-4b4a-93e4-5f571e91255a/oauth2/v2.0/token";
+            HttpClient client = new HttpClient();
+            client.BaseAddress = new Uri(url);
+            string clientAssertion = GetSignedClientAssertion();
+
+            var values = new Dictionary<string, string>
+            {
+                { "grant_type", "client_credentials" },
+                { "thing2", "e5e15768-1702-474d-ba7b-904c7cad2bcf" },
+                {"client_assertion", clientAssertion },
+                {"scope", "https://microsoft.sharepoint.com/.default" },
+                {"client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer" }
+            };
+
+            client.DefaultRequestHeaders.Add("Accept", "application/json");
+
+            var content = new FormUrlEncodedContent(values);
+            var response = await client.PostAsync("https://login.microsoftonline.com/f8cdef31-a31e-4b4a-93e4-5f571e91255a/oauth2/v2.0/token", content);
+            var responseString = await response.Content.ReadAsStringAsync();
+            JObject test = JObject.Parse(responseString);
+            return ((test["access_token"]).ToString());
+        }
+
+        // Code to fetch the Client Assertion from KeyVault
+        private string GetSignedClientAssertion()
+        {
+            X509Certificate2 selfSignedCertificate = new X509Certificate2(@"C:\Users\riagarwa\Downloads\oct5keyvault-ProdCert2-20201211.pfx", "", X509KeyStorageFlags.EphemeralKeySet);
```

Comment:
What would be a better choice relative or absolute path or some other source like config.